File: MaximumSubarray.java
/**
 * LeetCode 53 Maximum Subarray
 */
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int sum = 0;
        int max = Integer.MIN_VALUE;

        for (int i = 0; i < nums.length; i++) {
            if (sum + nums[i] > nums[i]) {
                sum += nums[i];
            } else {
                sum = nums[i];
            }
            if (sum > max) {
                max = sum;
            }
        }
        return max;
    }
}

File: MaximumSubarray.go
func maxSubArray(nums []int) int {
	if nums == nil || len(nums) == 0 {
		return 0
	}
	n := len(nums)
	dp := make([]int,n)
	dp[0] = nums[0]
	res := nums[0]
	for i := 1;i < n;i++ {
		dp[i] = max(dp[i - 1] + nums[i],nums[i])
		res = max(res,dp[i])
	}
	return res
}

func max(a int,b int) int {
	if (a > b) {
		return a
	}
	return b
}
